app/src/org/commcare/android/database/global/models/ApplicationRecord.java

```diff
@@ -56,4 +79,94 @@ public int getStatus() {
     public void setStatus(int status) {
         this.status = status;
     }
+
+    public String getUniqueId() {
+        return uniqueId;
+    }
+
+    public void setUniqueId(String id) {
+        this.uniqueId = id;
+    }
+
+    public String getDisplayName() {
+        return this.displayName;
+    }
+
+    public void setDisplayName(String appName) {
+        this.displayName = appName;
+    }
+
+    public void setArchiveStatus(boolean b) {
+        this.isArchived = b;
+    }
+
+    public boolean isArchived() {
+        return this.isArchived;
+    }
+
+    @Override
+    public String toString() {
+        return this.displayName;
+    }
+
+    public void setResourcesStatus(boolean b) {
+        this.resourcesValidated = b;
+    }
+
+    public boolean resourcesValidated() {
+        return this.resourcesValidated;
+    }
+
+    /*
+     * Returns true if this ApplicationRecord represents an app generated from
+     * a pre-Multiple Apps version of CommCare that does not have profile files
+     * with uniqueId and displayName
+     */
+    public boolean preMultipleAppsProfile() {
+        return this.preMultipleAppsProfile;
+    }
+    
+    public void setPreMultipleAppsProfile(boolean b) {
+        this.preMultipleAppsProfile = b;
+    }
+
+    /*
+     * Returns true if this ApplicationRecord was just generated from the a
+     * different ApplicationRecord format via the db upgrader, because it was 
+     * initially installed on a phone with a pre-Multiple Apps version of CommCare
+     */
+    public boolean convertedByDbUpgrader() {
+        return this.convertedViaDbUpgrader;
+    }
+
+    public void setConvertedByDbUpgrader(boolean b) {
+        this.convertedViaDbUpgrader = b;
+    }
+
+    /** Completes a full uninstall of the CommCare app that this ApplicationRecord represents **/
+    public void uninstall(Context c) {
+        CommCareApplication._().initializeAppResources(new CommCareApp(this));
+        CommCareApp app = CommCareApplication._().getCurrentApp();
+        
+        //1) Set states to delete requested so we know if we have left the app in a bad state later
+        CommCareApplication._().setAppResourceState(CommCareApplication.STATE_DELETE_REQUESTED);
+        this.setStatus(ApplicationRecord.STATUS_DELETE_REQUESTED);
+        CommCareApplication._().getGlobalStorage(ApplicationRecord.class).write(this);
+        //2) Teardown the sandbox for this app
+        app.teardownSandbox();   
+        //3) Delete all the user databases associated with this app
+        SqlStorage<UserKeyRecord> userDatabase = CommCareApplication._().
+                getAppStorage(UserKeyRecord.class);
+        for (UserKeyRecord user : userDatabase) {
+            c.getDatabasePath(CommCareUserOpenHelper.getDbName(user.getUuid())).delete();
+        }
+        //4) Delete the app database
+        c.getDatabasePath(DatabaseAppOpenHelper.getDbName(app.getAppRecord().

```

Comment:
I would recommend reading the return code of calls to 'delete' and logging instances where delete fails. Clayton disagrees with me :), but it is at least good for new code that we have yet to QA.